carp.common/src/commonMain/kotlin/dk/cachet/carp/common/application/URL.kt

```diff
@@ -0,0 +1,101 @@
+package dk.cachet.carp.common.application
+
+import dk.cachet.carp.common.infrastructure.serialization.createCarpStringPrimitiveSerializer
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.Serializable
+import kotlin.js.JsExport
+
+/**
+ * Represents a Uniform Resource Locator (URL) according to RFC 3986. We use [URIRegex] to match
+ * [stringRepresentation], but it cannot do actual validation. Therefore, [DefaultURLValidator] is used to validate
+ * the URL.
+ *
+ * The URL of the web page which contains the task to be performed.
+ * The URL may contain [URL.Variable] patterns which will be replaced with the corresponding values by the client
+ * runtime.
+ *
+ * @param stringRepresentation The string representation of the URL.
+ */
+@Serializable( URLSerializer::class )
+@JsExport
+class URL ( val stringRepresentation: String )
+{
+    init
+    {
+        require(
+            URIRegex.matches( stringRepresentation ) && DefaultURLValidator.isValid( stringRepresentation )
+        )
+        {
+            "Invalid URI string representation."
+        }
+    }
+
+    companion object
+    {
+        private fun markup( name: String ) = name.uppercase().replace( ' ', '-' ).let { "{{$it}}" }
+    }
+
+    fun replaceVariables( participantId: UUID, deploymentId: UUID, triggerId: Int ): String =
+            stringRepresentation.replace( Variable.PARTICIPANT_ID.pattern, participantId.toString() )
+                    .replace( Variable.DEPLOYMENT_ID.pattern, deploymentId.toString() )
+                    .replace( Variable.TRIGGER_ID.pattern, triggerId.toString() )
+
+    override fun equals( other: Any? ): Boolean
+    {
+        if ( this === other ) return true
+        if ( other !is URL ) return false
+
+        return stringRepresentation == other.stringRepresentation
+    }
+
+    override fun hashCode(): Int = stringRepresentation.hashCode()
+
+    override fun toString(): String = stringRepresentation
+
+    enum class Variable( val pattern: String )
+    {
+        /**
+         * Uniquely identifies the participant in the study.
+         */
+        PARTICIPANT_ID(markup("participant id")),
+
+        /**
+         * Uniquely identifies the deployment (group of participants and devices) of the study.
+         */
+        DEPLOYMENT_ID(markup("deployment id")),
+
+        /**
+         * Identifies the condition, defined by the study protocol, which causes the event to be triggered.
+         */
+        TRIGGER_ID(markup("trigger id"))
+    }
+}
+
+/**
+ * Regular expression to match any URI according to
+ * [Appendix B of RFC 3986/STD 0066](https://www.rfc-editor.org/rfc/rfc3986#appendix-B)
+ *
+ * TODO: [URIRegex] allows us to write our own validation, but for the most part, what we have now is sufficient.
```

Comment:
And, while looking deeper into this regex, there are two more issues:
1. This is not a regex to _validate_ the URL, but to extract the 5 components. The RFC you link explicitly calls it a "non-validating parser".
2. If it would actually be a regex to validate URLs, it would fail for expected input to this class, namely the URL variables. They are formatted as `{{VAR}}`, and `{` and `}` are not valid characters, which is exactly why they are a good pick for injecting variables. :)